File: voxel-world/src/java/com/systemvi/voxel/world/world/ChunkCache.java
package com.systemvi.voxel.world.world;

import com.systemvi.engine.model.Mesh;
import com.systemvi.engine.model.VertexAttribute;
import org.joml.Matrix4f;
import org.joml.Vector2f;
import org.joml.Vector3i;
import org.joml.Vector4f;

import java.util.ArrayList;

public class ChunkCache {
    public Mesh mesh;
    public World world;
    public Vector3i chunkPosition;
    public int triangles,instancesToDraw;
    public ChunkCache(World world,Vector3i chunkPosition){
        this.world=world;
        this.chunkPosition=chunkPosition;

        mesh=new Mesh(
            new VertexAttribute("position",2),
            new VertexAttribute("normal",3),
            new VertexAttribute("tangent",3),
            new VertexAttribute("bitangent",3),
            new VertexAttribute("texCoords",2)
        );
        float size=0.5f;
        mesh.setVertexData(new float[]{
            //position    //normal  //tangent  //bitangent //texCoords
            +size, +size, 0, 0, 1,  1, 0, 0,   0, 1, 0,    1, 0,
            -size, -size, 0, 0, 1,  1, 0, 0,   0, 1, 0,    0, 1,
            +size, -size, 0, 0, 1,  1, 0, 0,   0, 1, 0,    1, 1,
            -size, +size, 0, 0, 1,  1, 0, 0,   0, 1, 0,    0, 0,
        });
        mesh.setIndices(new int[]{
            0,2,1,
            0,1,3
        });
        mesh.enableInstancing(
            new VertexAttribute("col0",4),
            new VertexAttribute("col1",4),
            new VertexAttribute("col2",4),
            new VertexAttribute("col3",4),
            new VertexAttribute("uv",2),
            new VertexAttribute("lightLevel",4)
        );
        triangles=2;
        instancesToDraw=0;
        regenerate(world,chunkPosition);
    }
    public void regenerate(World world,Vector3i chunkPosition){
        ArrayList<Matrix4f> matrices=new ArrayList<>();
        ArrayList<Vector2f> uvs=new ArrayList<>();
        ArrayList<Vector4f> lightLevels=new ArrayList<>();
        Chunk chunk=world.getChunks()[chunkPosition.x][chunkPosition.y][chunkPosition.z];
        BlockState[][][] blockStates=chunk.blockStates;
        for(int i=0;i<Chunk.SIZE_X;i++){
            for(int j=0;j<Chunk.SIZE_Y;j++){
                for(int k=0;k<Chunk.SIZE_Z;k++){
                    if(blockStates[i][j][k].block== Block.AIR)continue;
                    int x=i+Chunk.SIZE_X*chunkPosition.x,y=j+Chunk.SIZE_Y*chunkPosition.y,z=k+Chunk.SIZE_Z*chunkPosition.z;
                    if(world.getBlockState(x-1,y,z).block== Block.AIR){
                        //left
                        matrices.add(new Matrix4f()
                            .translate(x,y,z)
                            .translate(-0.5f,0,0)
                            .rotateY((float)Math.toRadians(-90))
                        );
                        Block block=world.getBlockState(x,y,z).block;
                        uvs.add(new Vector2f(block.left.x,block.left.y));
                        lightLevels.add(new Vector4f(world.getBlockState(x-1,y,z).lightLevel));
                    }
                    if(world.getBlockState(x+1,y,z).block== Block.AIR){
                        //right
                        matrices.add(new Matrix4f()
                            .translate(x,y,z)
                            .translate(0.5f,0,0)
                            .rotateY((float)Math.toRadians(90))
                        );
                        Block block=world.getBlockState(x,y,z).block;
                        uvs.add(new Vector2f(block.right.x,block.right.y));
                        lightLevels.add(new Vector4f(world.getBlockState(x+1,y,z).lightLevel));
                    }
                    if(world.getBlockState(x,y-1,z).block== Block.AIR){
                        //down
                        matrices.add(new Matrix4f()
                            .translate(x,y,z)
                            .translate(0,-0.5f,0)
                            .rotateX((float)Math.toRadians(90))
                        );
                        Block block=world.getBlockState(x,y,z).block;
                        uvs.add(new Vector2f(block.bottom.x,block.bottom.y));
                        lightLevels.add(new Vector4f(world.getBlockState(x,y-1,z).lightLevel));
                    }
                    if(world.getBlockState(x,y+1,z).block== Block.AIR){
                        //up
                        matrices.add(new Matrix4f()
                            .translate(x,y,z)
                            .translate(0,0.5f,0)
                            .rotateX((float)Math.toRadians(-90))
                        );
                        Block block=world.getBlockState(x,y,z).block;
                        uvs.add(new Vector2f(block.top.x,block.top.y));
                        lightLevels.add(new Vector4f(world.getBlockState(x,y+1,z).lightLevel));
                    }
                    if(world.getBlockState(x,y,z-1).block== Block.AIR){
                        //back
                        matrices.add(new Matrix4f()
                            .translate(x,y,z)
                            .translate(0,0,-0.5f)
                            .rotateY((float)Math.toRadians(180))
                        );
                        Block block=world.getBlockState(x,y,z).block;
                        uvs.add(new Vector2f(block.back.x,block.back.y));
                        lightLevels.add(new Vector4f(world.getBlockState(x,y,z-1).lightLevel));
                    }
                    if(world.getBlockState(x,y,z+1).block== Block.AIR){
                        //front
                        matrices.add(new Matrix4f()
                            .translate(x,y,z)
                            .translate(0,0,0.5f)
                            .rotateY((float)Math.toRadians(0))
                        );
                        Block block=world.getBlockState(x,y,z).block;
                        uvs.add(new Vector2f(block.front.x,block.front.y));
                        lightLevels.add(new Vector4f(world.getBlockState(x,y,z+1).lightLevel));
                    }
                }
            }
        }
        int matrixSize=16;
        int uvSize=2;
        int lightLevelSize=4;
        int instanceSize=matrixSize+uvSize+lightLevelSize;
        float[] instanceData=new float[matrices.size()*instanceSize];
        float[] matrixData=new float[matrixSize];
        float[] data=new float[uvSize];
        for(int i=0;i<matrices.size();i++){
            matrices.get(i).get(matrixData);
            data[0]=uvs.get(i).x;
            data[1]=uvs.get(i).y;
            for(int j=0;j<matrixSize;j++){
                instanceData[i*instanceSize+j]=matrixData[j];
            }
            for(int j=0;j<uvSize;j++){
                instanceData[i*instanceSize+matrixSize+j]=data[j];
            }

        }
        instancesToDraw=matrices.size();
        mesh.setInstanceData(instanceData);
    }
}

File: voxel-world/src/scala/com/systemvi/voxel/world/world2/ChunkCache.scala
package com.systemvi.voxel.world.world2

import com.systemvi.voxel.world.world2
import org.joml.{Vector3i, Vector4f}

class ChunkCache(world: World, chunk: Chunk, chunkPosition: Vector3i) {

  var blockFaces: List[BlockFace] = List.empty
  private val w = Vector3i(
    chunkPosition.x * Chunk.size.x,
    chunkPosition.y * Chunk.size.y,
    chunkPosition.z * Chunk.size.z
  )
  for (i <- 0 until Chunk.size.x; j <- 0 until Chunk.size.y; k <- 0 until Chunk.size.z) {
    val occlusions = Array(1f, 0.8f, 0.5f, 0f)
    val worldPosition = Vector3i(w).add(i, j, k)
    chunk.blockStates(i)(j)(k) match
      case state: BlockState if state.block == Block.AIR =>
      case state: BlockState =>
        //add left side
        if world.getBlockState(worldPosition.x - 1, worldPosition.y, worldPosition.z).block == Block.AIR then
          val top = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z)
          val bottom = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z)
          val front = world.getBlockState(worldPosition.x - 1, worldPosition.y, worldPosition.z + 1)
          val back = world.getBlockState(worldPosition.x - 1, worldPosition.y, worldPosition.z - 1)
          val topFront = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z + 1)
          val topBack = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z - 1)
          val bottomFront = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z + 1)
          val bottomBack = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z - 1)
          val x = Array(bottom, back, bottomBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val y = Array(bottom, front, bottomFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val z = Array(top, back, topBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val w = Array(top, front, topFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val occlusion = Vector4f(occlusions(x), occlusions(y), occlusions(z), occlusions(w))
          blockFaces :+= BlockFace(state, worldPosition, occlusion, BlockSide.Left)
        //add right side
        if world.getBlockState(worldPosition.x + 1, worldPosition.y, worldPosition.z).block == Block.AIR then
          val top = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z)
          val bottom = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z)
          val front = world.getBlockState(worldPosition.x + 1, worldPosition.y, worldPosition.z + 1)
          val back = world.getBlockState(worldPosition.x + 1, worldPosition.y, worldPosition.z - 1)
          val topFront = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z + 1)
          val topBack = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z - 1)
          val bottomFront = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z + 1)
          val bottomBack = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z - 1)
          val x = Array(bottom, front, bottomFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val y = Array(bottom, back, bottomBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val z = Array(top, front, topFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val w = Array(top, back, topBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val occlusion = Vector4f(occlusions(x), occlusions(y), occlusions(z), occlusions(w))
          //          val occlusion = Vector4f(occlusions(x),1,1,1)
          blockFaces :+= BlockFace(state, worldPosition, occlusion, BlockSide.Right)
        //add bottom side
        if world.getBlockState(worldPosition.x, worldPosition.y - 1, worldPosition.z).block == Block.AIR then
          val left = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z)
          val right = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z)
          val front = world.getBlockState(worldPosition.x, worldPosition.y - 1, worldPosition.z + 1)
          val back = world.getBlockState(worldPosition.x, worldPosition.y - 1, worldPosition.z - 1)
          val leftFront = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z + 1)
          val leftBack = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z - 1)
          val rightFront = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z + 1)
          val rightBack = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z - 1)
          val x = Array(left, front, leftFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val y = Array(right, front, rightFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val z = Array(left, back, leftBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val w = Array(right, back, rightBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val occlusion = Vector4f(occlusions(x), occlusions(y), occlusions(z), occlusions(w))
          //          val occlusion = Vector4f(0,1,1,1),1,1)
          blockFaces :+= BlockFace(state, worldPosition, occlusion, BlockSide.Bottom)
        //add top side
        if world.getBlockState(worldPosition.x, worldPosition.y + 1, worldPosition.z).block == Block.AIR then
          val left = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z)
          val right = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z)
          val front = world.getBlockState(worldPosition.x, worldPosition.y + 1, worldPosition.z + 1)
          val back = world.getBlockState(worldPosition.x, worldPosition.y + 1, worldPosition.z - 1)
          val leftFront = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z + 1)
          val leftBack = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z - 1)
          val rightFront = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z + 1)
          val rightBack = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z - 1)
          val x = Array(left, front, leftFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val y = Array(right, front, rightFront).map(state => if state.block != Block.AIR then 1 else 0).sum
          val z = Array(left, back, leftBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val w = Array(right, back, rightBack).map(state => if state.block != Block.AIR then 1 else 0).sum
          val occlusion = Vector4f(occlusions(x), occlusions(y), occlusions(z), occlusions(w))
//          val occlusion = Vector4f(0,1,1,1)
          blockFaces :+= BlockFace(state, worldPosition, occlusion, BlockSide.Top)
        //add front
        if world.getBlockState(worldPosition.x, worldPosition.y, worldPosition.z + 1).block == Block.AIR then
          val left        = world.getBlockState(worldPosition.x - 1, worldPosition.y, worldPosition.z + 1)
          val right       = world.getBlockState(worldPosition.x + 1, worldPosition.y , worldPosition.z + 1)
          val top         = world.getBlockState(worldPosition.x, worldPosition.y + 1, worldPosition.z + 1)
          val bottom      = world.getBlockState(worldPosition.x, worldPosition.y - 1, worldPosition.z + 1)
          val topLeft     = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z + 1)
          val topRight    = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z + 1)
          val bottomLeft  = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z + 1)
          val bottomRight = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z + 1)
          val x = Array(left, bottom, bottomLeft).map(state => if state.block != Block.AIR then 1 else 0).sum
          val y = Array(right, bottom, bottomRight).map(state => if state.block != Block.AIR then 1 else 0).sum
          val z = Array(left, top, topLeft).map(state => if state.block != Block.AIR then 1 else 0).sum
          val w = Array(right, top, topRight).map(state => if state.block != Block.AIR then 1 else 0).sum
          val occlusion = Vector4f(occlusions(x), occlusions(y), occlusions(z), occlusions(w))
//          val occlusion = Vector4f(0,1,1,1)
          blockFaces :+= BlockFace(state, worldPosition, occlusion, BlockSide.Front)
        //add back
        if world.getBlockState(worldPosition.x, worldPosition.y, worldPosition.z - 1).block == Block.AIR then
          val left        = world.getBlockState(worldPosition.x - 1, worldPosition.y, worldPosition.z - 1)
          val right       = world.getBlockState(worldPosition.x + 1, worldPosition.y , worldPosition.z - 1)
          val top         = world.getBlockState(worldPosition.x, worldPosition.y + 1, worldPosition.z - 1)
          val bottom      = world.getBlockState(worldPosition.x, worldPosition.y - 1, worldPosition.z - 1)
          val topLeft     = world.getBlockState(worldPosition.x - 1, worldPosition.y + 1, worldPosition.z - 1)
          val topRight    = world.getBlockState(worldPosition.x + 1, worldPosition.y + 1, worldPosition.z - 1)
          val bottomLeft  = world.getBlockState(worldPosition.x - 1, worldPosition.y - 1, worldPosition.z - 1)
          val bottomRight = world.getBlockState(worldPosition.x + 1, worldPosition.y - 1, worldPosition.z - 1)
          val x = Array(right, bottom, bottomRight).map(state => if state.block != Block.AIR then 1 else 0).sum
          val y = Array(left, bottom, bottomLeft).map(state => if state.block != Block.AIR then 1 else 0).sum
          val z = Array(right, top, topRight).map(state => if state.block != Block.AIR then 1 else 0).sum
          val w = Array(left, top, topLeft).map(state => if state.block != Block.AIR then 1 else 0).sum
          val occlusion = Vector4f(occlusions(x), occlusions(y), occlusions(z), occlusions(w))
          //          val occlusion = Vector4f(0,1,1,1)
          blockFaces :+= BlockFace(state, worldPosition, occlusion, BlockSide.Back)
  }
}
